File: GithubGists/app/src/main/java/ru/avystavkin/githubgists/repository/github/DefaultGithubRepository.java
package ru.avystavkin.githubgists.repository.github;

import java.util.List;

import androidx.annotation.NonNull;
import io.reactivex.Observable;
import ru.avystavkin.githubgists.api.GithubService;
import ru.avystavkin.githubgists.database.DbHelper;
import ru.avystavkin.githubgists.models.local.Gist;
import ru.avystavkin.githubgists.models.local.GistCommit;
import ru.avystavkin.githubgists.models.local.User;
import ru.avystavkin.githubgists.repository.github.transformer.RxGistCommitTransformer;
import ru.avystavkin.githubgists.repository.github.transformer.RxGistTransformer;
import ru.avystavkin.githubgists.repository.github.transformer.RxGistsListTransformer;
import ru.avystavkin.githubgists.utils.RxUtils;

public class DefaultGithubRepository implements GithubRepository {

    private final GithubService mGithubService;
    private final DbHelper mDbHelper;

    public DefaultGithubRepository(
            @NonNull DbHelper dbHelper,
            @NonNull GithubService githubService) {
        mGithubService = githubService;
        mDbHelper = dbHelper;
    }

    @NonNull
    @Override
    public Observable<List<Gist>> getGists() {
         return mGithubService.gists()
                 .compose(new RxGistsListTransformer())
                 .flatMap(gists -> {
                     mDbHelper.insert(gists);
                     return Observable.fromArray(gists);
                 })
                 .onErrorReturn(throwable -> mDbHelper.getGists())
                 .compose(RxUtils.async());
    }

    @NonNull
    @Override
    public Observable<List<User>> getPopularUsers(int count) {
        return  Observable.just(true)
                .flatMap(x -> Observable.fromArray(mDbHelper.getPopularUsers(count)))
                .compose(RxUtils.async());
    }

    @NonNull
    @Override
    public Observable<Gist> getGistById(String id) {
        return mGithubService.gist_detail(id)
                .compose(new RxGistTransformer())
                .compose(RxUtils.async());
    }

    @NonNull
    @Override
    public Observable<List<Gist>> getGistsByUserName(String name) {
        return mGithubService.user_detail(name)
                .compose(new RxGistsListTransformer())
                .compose(RxUtils.async());
    }

    @NonNull
    @Override
    public Observable<List<GistCommit>> getCommitsByGistId(String id) {
        return mGithubService.gist_commits(id)
                .compose(new RxGistCommitTransformer())
                .compose(RxUtils.async());
    }
}

File: GithubGists/app/src/main/java/ru/avystavkin/githubgists/repository/github/DefaultGithubRepository.kt
package ru.avystavkin.githubgists.repository.github

import io.reactivex.Observable
import ru.avystavkin.githubgists.api.GithubService
import ru.avystavkin.githubgists.database.DbHelper
import ru.avystavkin.githubgists.models.local.Gist
import ru.avystavkin.githubgists.models.local.GistCommit
import ru.avystavkin.githubgists.models.local.User
import ru.avystavkin.githubgists.repository.github.transformer.RxGistCommitTransformer
import ru.avystavkin.githubgists.repository.github.transformer.RxGistTransformer
import ru.avystavkin.githubgists.repository.github.transformer.RxGistsListTransformer
import ru.avystavkin.githubgists.utils.RxUtils

class DefaultGithubRepository(
        private val mDbHelper: DbHelper,
        private val mGithubService: GithubService) : GithubRepository {

    override val gists: Observable<List<Gist>>
        get() = mGithubService.gists()
                .compose(RxGistsListTransformer())
                .flatMap { gists ->
                    mDbHelper.insert(gists)
                    Observable.fromArray(gists)
                }
                .onErrorReturn { mDbHelper.getGists() }
                .compose(RxUtils.async())

    override fun getPopularUsers(count: Int): Observable<List<User>> {
        return Observable.just(true)
                .flatMap { x -> Observable.fromArray(mDbHelper.getPopularUsers(count)) }
                .compose(RxUtils.async())
    }

    override fun getGistById(id: String): Observable<Gist> {
        return mGithubService.gist_detail(id)
                .compose(RxGistTransformer())
                .compose(RxUtils.async())
    }

    override fun getGistsByUserName(name: String): Observable<List<Gist>> {
        return mGithubService.user_detail(name)
                .compose(RxGistsListTransformer())
                .compose(RxUtils.async())
    }

    override fun getCommitsByGistId(id: String): Observable<List<GistCommit>> {
        return mGithubService.gist_commits(id)
                .compose(RxGistCommitTransformer())
                .compose(RxUtils.async())
    }
}
